File: backend/src/server_config.py
#!/usr/bin/env python3
"""
Backend Server Configuration Management Utility
Helps switch between async and sync backend servers
"""

import os
import sys
import subprocess
import requests
import json
from pathlib import Path
from typing import Dict, Optional, Tuple

# Server configurations
ASYNC_CONFIG = {
    "name": "ASYNC",
    "port": 8001,
    "host": "0.0.0.0",
    "main_file": "main_async:app",
    "features": ["Redis Caching", "Async I/O", "Performance Monitoring", "Enhanced Logging"],
    "endpoints": 43,
    "test_credentials": "superadmin/superadmin123",
    "env_file": ".env.async",
    "database": "spendplatform_async.db"
}

SYNC_CONFIG = {
    "name": "SYNC",
    "port": 8000,
    "host": "0.0.0.0",
    "main_file": "main:app",
    "features": ["Traditional Sync I/O", "Standard Endpoints"],
    "endpoints": 20,
    "test_credentials": "user/user123 or superadmin/superadmin123",
    "env_file": ".env.sync",
    "database": "spendplatform_sync.db"
}

def get_project_root() -> Path:
    """Get the project root directory"""
    current = Path(__file__).parent
    while current.parent != current:
        if (current / "start_backend.sh").exists():
            return current
        current = current.parent
    return Path.cwd()

def check_server_health(port: int) -> Tuple[bool, Optional[Dict]]:
    """Check if server is running and healthy"""
    try:
        response = requests.get(f"http://localhost:{port}/health", timeout=5)
        if response.status_code == 200:
            return True, response.json()
        return False, None
    except Exception as e:
        return False, {"error": str(e)}

def get_current_server_type() -> str:
    """Determine current server type from environment"""
    env_file = get_project_root() / "backend" / ".env"
    if env_file.exists():
        with open(env_file, 'r') as f:
            content = f.read()
            if "SERVER_TYPE=ASYNC" in content:
                return "ASYNC"
            elif "SERVER_TYPE=SYNC" in content:
                return "SYNC"
    return "UNKNOWN"

def switch_to_async():
    """Switch backend to async server configuration"""
    project_root = get_project_root()
    backend_dir = project_root / "backend"
    
    print("🔄 Switching to ASYNC backend server...")
    
    # Copy async environment
    env_async = backend_dir / ".env.async"
    env_target = backend_dir / ".env"
    
    if env_async.exists():
        subprocess.run(["cp", str(env_async), str(env_target)])
        print("✅ Environment configuration updated")
    else:
        print("❌ .env.async file not found")
        return False
    
    print("⚡ ASYNC server configuration active")
    print(f"🌐 Server will run on: http://localhost:{ASYNC_CONFIG['port']}")
    print(f"🔐 Test credentials: {ASYNC_CONFIG['test_credentials']}")
    print(f"🚀 Features: {', '.join(ASYNC_CONFIG['features'])}")
    print("")
    print("To start the server, run:")
    print("  ./start_backend_async.sh")
    print("  # OR")
    print("  ./start_backend.sh  # (now defaults to async)")
    
    return True

def switch_to_sync():
    """Switch backend to sync server configuration"""
    project_root = get_project_root()
    backend_dir = project_root / "backend"
    
    print("🔄 Switching to SYNC backend server...")
    
    # Copy sync environment
    env_sync = backend_dir / ".env.sync"
    env_target = backend_dir / ".env"
    
    if env_sync.exists():
        subprocess.run(["cp", str(env_sync), str(env_target)])
        print("✅ Environment configuration updated")
    else:
        print("❌ .env.sync file not found")
        return False
    
    print("🐌 SYNC server configuration active")
    print(f"🌐 Server will run on: http://localhost:{SYNC_CONFIG['port']}")
    print(f"🔐 Test credentials: {SYNC_CONFIG['test_credentials']}")
    print(f"🚀 Features: {', '.join(SYNC_CONFIG['features'])}")
    print("")
    print("To start the server, run:")
    print("  ./start_backend_sync.sh")
    print("  # OR")
    print("  BACKEND_SERVER_TYPE=sync ./start_backend.sh")
    
    return True

def display_current_config():
    """Display current backend configuration"""
    current_type = get_current_server_type()
    config = ASYNC_CONFIG if current_type == "ASYNC" else SYNC_CONFIG
    
    print("🔧 Current Backend Configuration:")
    print(f"   Server Type: {config['name']}")
    print(f"   Port: {config['port']}")
    print(f"   Host: {config['host']}")
    print(f"   Main File: {config['main_file']}")
    print(f"   Database: {config['database']}")
    print(f"   Features: {', '.join(config['features'])}")
    print(f"   Endpoints: {config['endpoints']}+")
    print(f"   Test Credentials: {config['test_credentials']}")
    print(f"   Environment File: {config['env_file']}")
    print("")
    
    return config

def check_both_servers():
    """Check the health of both servers"""
    print("🔍 Checking server health...")
    
    # Check async server
    async_healthy, async_data = check_server_health(ASYNC_CONFIG['port'])
    print(f"⚡ ASYNC Server (port {ASYNC_CONFIG['port']}): {'✅ HEALTHY' if async_healthy else '❌ NOT RUNNING'}")
    if async_healthy and async_data:
        print(f"   Services: {async_data.get('services', 'Unknown')}")
        print(f"   Version: {async_data.get('version', 'Unknown')}")
    
    # Check sync server
    sync_healthy, sync_data = check_server_health(SYNC_CONFIG['port'])
    print(f"🐌 SYNC Server (port {SYNC_CONFIG['port']}): {'✅ HEALTHY' if sync_healthy else '❌ NOT RUNNING'}")
    if sync_healthy and sync_data:
        print(f"   Status: {sync_data.get('status', 'Unknown')}")
    
    print("")
    return {
        "async": {"healthy": async_healthy, "data": async_data},
        "sync": {"healthy": sync_healthy, "data": sync_data}
    }

def main():
    """Main CLI interface"""
    if len(sys.argv) < 2:
        print("🔧 Backend Server Configuration Manager")
        print("")
        print("Usage:")
        print("  python server_config.py status     # Show current configuration")
        print("  python server_config.py async      # Switch to async server")
        print("  python server_config.py sync       # Switch to sync server")
        print("  python server_config.py health     # Check both servers")
        print("")
        display_current_config()
        return
    
    command = sys.argv[1].lower()
    
    if command == "status":
        display_current_config()
    elif command == "async":
        switch_to_async()
    elif command == "sync":
        switch_to_sync()
    elif command == "health":
        check_both_servers()
    else:
        print(f"❌ Unknown command: {command}")
        print("Available commands: status, async, sync, health")

if __name__ == "__main__":
    main()

File: frontend/src/services/serverConfig.ts
// Frontend-Backend Server Connection Configuration Manager
// This utility helps switch between async and sync backend servers

export interface ServerConfig {
  baseURL: string;
  serverType: 'ASYNC' | 'SYNC';
  timeout: number;
  apiPrefix: string;
  testCredentials: {
    username: string;
    password: string;
  };
  features: {
    redis: boolean;
    caching: boolean;
    performanceHeaders: boolean;
  };
}

// ASYNC Server Configuration (Default - High Performance)
export const ASYNC_CONFIG: ServerConfig = {
  baseURL: process.env.REACT_APP_API_BASE_URL || 'http://localhost:8001',
  serverType: 'ASYNC',
  timeout: parseInt(process.env.REACT_APP_API_TIMEOUT || '10000'),
  apiPrefix: '/api/v1',
  testCredentials: {
    username: process.env.REACT_APP_DEFAULT_TEST_USER || 'superadmin',
    password: process.env.REACT_APP_DEFAULT_TEST_PASSWORD || 'superadmin123'
  },
  features: {
    redis: true,
    caching: true,
    performanceHeaders: true
  }
};

// SYNC Server Configuration (Backup - Traditional)
export const SYNC_CONFIG: ServerConfig = {
  baseURL: 'http://localhost:8000',
  serverType: 'SYNC',
  timeout: 5000,
  apiPrefix: '',
  testCredentials: {
    username: 'user',
    password: 'user123'
  },
  features: {
    redis: false,
    caching: false,
    performanceHeaders: false
  }
};

// Get current server configuration based on environment
export const getCurrentConfig = (): ServerConfig => {
  const serverType = process.env.REACT_APP_SERVER_TYPE || 'ASYNC';
  
  if (serverType === 'SYNC') {
    return SYNC_CONFIG;
  }
  
  return ASYNC_CONFIG;
};

// Get API endpoint with proper prefix
export const getApiEndpoint = (endpoint: string): string => {
  const config = getCurrentConfig();
  
  // Remove leading slash if present
  const cleanEndpoint = endpoint.startsWith('/') ? endpoint.slice(1) : endpoint;
  
  // Add prefix for async server, no prefix for sync
  if (config.serverType === 'ASYNC') {
    return `${config.apiPrefix}/${cleanEndpoint}`;
  }
  
  return `/${cleanEndpoint}`;
};

// Server switching utilities
export const switchToAsyncServer = () => {
  console.log('🔄 Switching to ASYNC server (port 8001)...');
  console.log('⚡ Features: Redis caching, performance headers, enhanced async operations');
  console.log('🔐 Test credentials: superadmin/superadmin123');
  
  // In a real implementation, this would update environment variables
  // For now, it logs instructions
  console.log('📝 To switch permanently:');
  console.log('   1. Copy .env.async to .env');
  console.log('   2. Restart the React development server');
  console.log('   3. Update imports to use api-async.ts instead of api.ts if needed');
};

export const switchToSyncServer = () => {
  console.log('🔄 Switching to SYNC server (port 8000)...');
  console.log('🐌 Features: Traditional sync operations, no caching');
  console.log('🔐 Test credentials: user/user123');
  
  console.log('📝 To switch permanently:');
  console.log('   1. Copy .env.sync to .env');
  console.log('   2. Restart the React development server');
  console.log('   3. Update imports to use api-sync-backup.ts instead of api.ts if needed');
};

// Health check for both servers
export const checkBothServers = async () => {
  const results = {
    async: false,
    sync: false,
    asyncError: null as any,
    syncError: null as any
  };

  // Check async server
  try {
    const asyncResponse = await fetch('http://localhost:8001/health');
    results.async = asyncResponse.ok;
  } catch (error) {
    results.asyncError = error;
  }

  // Check sync server
  try {
    const syncResponse = await fetch('http://localhost:8000/health');
    results.sync = syncResponse.ok;
  } catch (error) {
    results.syncError = error;
  }

  return results;
};

// Display current configuration
export const displayCurrentConfig = () => {
  const config = getCurrentConfig();
  
  console.log('🔧 Current Frontend Configuration:');
  console.log(`   Server Type: ${config.serverType}`);
  console.log(`   Base URL: ${config.baseURL}`);
  console.log(`   API Prefix: ${config.apiPrefix || 'None'}`);
  console.log(`   Timeout: ${config.timeout}ms`);
  console.log(`   Redis Caching: ${config.features.redis ? 'Enabled' : 'Disabled'}`);
  console.log(`   Performance Headers: ${config.features.performanceHeaders ? 'Enabled' : 'Disabled'}`);
  console.log(`   Test User: ${config.testCredentials.username}`);
  
  return config;
};

export default {
  ASYNC_CONFIG,
  SYNC_CONFIG,
  getCurrentConfig,
  getApiEndpoint,
  switchToAsyncServer,
  switchToSyncServer,
  checkBothServers,
  displayCurrentConfig
};